File: ingest/pkg/model/weather.go
package model

type Forecast struct {
    Source                   string  `json:"source"`

    Time                     int     `json:"time"`
    TimeForecasted           int     `json:"time_forecasted"`

    Latitude                 float64 `json:"latitude"`
    Longitude                float64 `json:"longitude"`
    City                     string  `json:"city"`
    Country                  string  `json:"country"`

    Temperature              float64 `json:"temperature"`
    TemperatureFeelsLike     float64 `json:"temperature_feels_like"`

    Pressure                 int     `json:"pressure"`
    Humidity                 float64 `json:"humidity"`
    PrecipitationProbability float64 `json:"precipitation_probability"`
    Cloudiness               float64 `json:"cloudiness"`
    Visibility               float64 `json:"visibility"`
    RainVolume               float64 `json:"rain_volume"`
    SnowVolume               float64 `json:"snow_volume"`
    WindSpeed                float64 `json:"wind_speed"`
    WindAngle                int     `json:"wind_angle"`
    WindGust                 float64 `json:"wind_gust"`
}

File: predict/model/dataset/weather.py
import torch

import numpy as np
import pandas as pd

from google.cloud import bigquery

class Weather (torch.utils.data.Dataset):
    def __init__ (self):
        # connect to bigquery
        client = bigquery.Client()

        # load dataset
        data = client.query("SELECT * from weather")
        data = data.to_dataframe()

        # store dataset and tensor column selection
        self.data = data
        self.tensor_columns = [
            "latitude_sin", "latitude_cos",
            "longitude_sin", "longitude_cos",
            "week_of_the_year_sin", "week_of_the_year_cos",
            "day_of_the_week_sin", "day_of_the_week_cos",
            "temperature", "temperature_feels_like",
            "pressure", "humidity", "cloudiness", "visibility",
            "precipitation_probability", "rain_volume", "snow_volume",
            "wind_speed", "wind_angle", "wind_speed"
        ]

    # override len(data)
    def __len__ (self):
        return len(self.data)

    # override data[idx]
    def __getitem__ (self, idx):

        location = torch.tensor(self.data.iloc[idx][self.tensor_columns[:8]].values)
        history = self.locationHistory(forecast["latitude"], forecast["longitude"], 100, forecast["time"], forecast["time"] - 60*60*24*7)
        forecast = torch.tensor(self.data.iloc[idx][self.tensor_columns[8:]].values)

        return location, history, forecast

    # weather history for a location + radius
    def locationHistory (self, latitude, longitude, radius, time_start, time_stop):
        radius_earth = 6378

        latitude_offset = (radius / radius_earth) * (180 / np.pi)
        longitude_offset = (radius / radius_earth) * (180 / np.pi) / np.cos(latitude * np.pi / 180)

        latitude_filter = self.data["latitude"].between(latitude-latitude_offset, latitude+latitude_offset)
        longitude_filter = self.data["longitude"].between(longitude-longitude_offset, longitude+longitude_offset)

        history = self.data[latitude_filter and longitude_filter and self.data["time"].between(time_start, time_stop)]
        history = history[self.tensor_columns]

        return torch.Tensor(history.values)
